### ISSUE TITLE
Clippy incorrectly complains about `borrowed_box` when trait object is boxed

### ISSUE THREAD

**Issue by idubrov** [rust-lang/rust-clippy#3971]
The following code triggers `clippy::borrowed_box`:

```rust
trait Whatever {}
struct Something;
impl Whatever for Something {}
fn func(flag: bool, secret: &mut Box<Whatever>) {
    *secret = Box::new(Something);
}
```

However, I don't think it should be complaining here since `Whatever` is not `Sized` (so I cannot overwrite it directly)

I think, this issue (or similar) was raised before: #1884 (also mentioned in #3128)

The fix for #1884 was, however, just to check for `Any` trait as a special case.

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5c0f61a8d819c01535527bbdb6faf748)

**Comment by flip1995** [comment]
borrowed_box doesn't trigger on the code anymore.

**Comment by MarcelCoding** [comment]
It is doing it again. I updated the example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f1d0ac1aa9020c047580ac51061f1af5

**Comment by flip1995** [comment]
What prevents you from just using `&dyn Whatever` in this example? According to the [lint documentation](https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box) this is exactly the case this lint was made for.

**Comment by MarcelCoding** [comment]
Whatever is a trait an at the point without a reference I have to box the trait.

**Comment by xanathar** [comment]
Can you do like `this_is_fine` in this example?

```rust
trait Whatever {}
struct Something;
impl Whatever for Something {}

fn this_is_fine(_: &dyn Whatever) { }

fn this_is_also_fine_but_moves(_: Box<dyn Whatever>) { }

fn this_triggers_clippy(_: &Box<dyn Whatever>) { }


fn main() {
    let wt = Box::new(Something);
    let wt: Box<dyn Whatever> = wt;

    this_triggers_clippy(&wt);
    this_is_fine(&*wt);
    this_is_also_fine_but_moves(wt);
}
```

**Comment by flip1995** [comment]
Alternatively `this_is_fine(wt.as_ref())`. The advantage of having it as just `&dyn _` is that if you have a concrete type that you want to pass to the function, you don't have to box it. So with `this_is_fine` you can just call `this_is_fine(&Something)`. But with `this_triggers_clippy` you have to box `Something` first: `this_triggers_clippy(&Box::new(Something))`. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8802668dbf0222fb004cff61bf1049f9)

**Comment by MarcelCoding** [comment]
Thanks, with the help of your example I got it working. I think it would be helpful for others to include this conversation in the message provided by clippy.

**Comment by flip1995** [comment]
Hard to include all of this in the message, but we should improve the documentation a bit.

**Comment by MarcelCoding** [comment]
Ether that, I rather thought about a link to this conversation.
